examples/linear_deformation.py

```diff
@@ -0,0 +1,1187 @@
+#        return (vectorial_ft_op_inverse(ft_displacement) /
+#                self.par_space[0].cell_volume * 2.0 * np.pi)
+
+
+class L2DataMatchingFunctional(Functional):
+
+    """Basic data matching functional using the L2 norm.
+
+    This functional computes::
+
+        1/2 * ||f - g||_2^2
+
+    for a given element ``g``.
+    """
+
+    def __init__(self, space, data):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        space : `DiscreteLp` with exponent 2.0
+            Space where the data is matched
+        data : `DiscreteLp` element-like
+            Data which is to be matched
+        """
+        if not (isinstance(space, odl.DiscreteLp) and space.exponent == 2.0):
+            raise ValueError('not an L2 space.')
+        super().__init__(space, linear=False)
+        self.data = self.domain.element(data)
+
+    def _call(self, x):
+        """Return ``self(x)``."""
+        return self.domain.dist(x, self.data)
+
+    def gradient(self, x):
+        """Return the gradient in the point ``x``."""
+        return x - self.data
+
+    def derivative(self, x):
+        """Return the derivative in ``x``."""
+        return self.gradient(x).T
+
+
+# Kernel function for any dimensional
+def gauss_kernel(x, sigma):
+    return np.exp(-x ** 2 / (2 * sigma ** 2))
+
+
+# Kernel function
+def kernel(x):
+    scaled = [xi ** 2 / (2 * sigma ** 2) for xi in x]
+    return np.exp(-sum(scaled))
+
+
+def gaussian_kernel_matrix(grid, sigma):
+    """Return the kernel matrix for Gaussian kernel.
+
+    The Gaussian kernel matrix ``K`` in ``n`` dimensions is defined as::
+
+        k_ij = exp(- |x_i - x_j|^2 / (2 * sigma^2))
+
+    where ``x_i, x_j`` runs through all grid points. The final matrix
+    has size ``N x N``, where ``N`` is the total number of grid points.
+
+    Parameters
+    ----------
+    grid : `RegularGrid`
+        Grid where the control points are defined
+    sigma : `float`
+        Width of the Gaussian kernel
+    """
+    point_arrs = grid.points().T
+    matrices = [parr[:, None] - parr[None, :] for parr in point_arrs]
+    for mat in matrices:
+        mat *= mat
+
+    sq_sum = np.sqrt(np.sum(mat for mat in matrices))
+    kernel_matrix = np.exp(-sq_sum / (2 * sigma ** 2))
+    return kernel_matrix
+
+
+def proj_noise(proj_data_shape, mu=0.0, sigma=0.1):
+    """Produce white Gaussian noise for projections of n-D images.
+
+       Produce white Gaussian noise for projections, with the same size
+       as the number of projections.
+
+       Parameters
+       ----------
+       proj_data_shape : shape of the projections
+           Give the size of noise
+       mu : Mean of the norm distribution
+           The defalt is 0.0
+       sigma : Standard deviation of the norm distribution
+           The defalt is 0.1.
+    """
+
+    return np.random.normal(mu, sigma, proj_data_shape)
+
+
+def SNR(signal, noise):
```

Comment:
The SciPy function has been deprecated, and I don't find a replacement. Anyway, the function is useful, and it should go into `odl.util`, perhaps in `numerics.py`?
Anyway, check out [this Wikipedia article](https://en.wikipedia.org/wiki/Signal-to-noise_ratio_(imaging)) for a reasonable definition. I think SNR should be defined either in terms of a single image or in terms of a signal image and a noise image, and the function should handle both cases, to have more flexibility. Also it should be able to compute in linear scale and logarithmic scale, perhaps even with different exponents? Should not be hard:
```python
scale, scale_in = str(scale).lower(), scale
if scale.startswith('log'):
    exponent = float(scale[3:])
    scale = 'log'
elif str(scale) == 'db':
    exponent = 10.0
    scale = 'log'

# Choice: no noise -> perfect signal -> ratio inf
if noise_power == 0.0:
    return float('inf')

if scale='linear':
    return signal_power / noise_power
elif scale == 'log':
    return exponent * np.log(signal_power / noise_power) / np.log(exponent)
else:
    # Error
```